src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs

```diff
@@ -755,7 +755,16 @@ private bool TryTakeWithNoTimeValidation([MaybeNullWhen(false)] out T item, int
                     }
                 }
             }
+
+            if (waitForSemaphoreWasSuccessful)
+            {
+                Debug.Assert(item != null);
+            }
+
+#pragma warning disable CS8762
+            // Compiler can't automatically deduce 'item' has a non-null value when returning false.
             return waitForSemaphoreWasSuccessful;
+#pragma warning restore CS8762
```

Comment:
There's no way to suppress this warning with a well-placed `!` somewhere?  The clutter caused by `!` is bad enough, but adding a bunch of pragmas to suppress false positive nullable warnings just makes me sad.
cc: @jcouv

Reply:
It's still not great, but possibly a bit better: add a `return true;` below the `Debug.Assert(item != null);` at line 761.

From the language/compiler perspective, it may be useful to allow `return waitForSemaphoreWasSuccessful!;` to solve this.

---
In reply to: [422141700](https://github.com/dotnet/runtime/pull/36104#discussion_r422141700) [](ancestors = 422141700)

Reply:
> From the language/compiler perspective, it may be useful to allow return waitForSemaphoreWasSuccessful!; to solve this.

@jcouv, yeah, let's do that :smile:

Reply:
I did consider using the `if (condition) { assert(); return true; } else { return false; }` pattern, but it produces different IL and codegen than a simple `return condition;`. I tried to avoid making changes that affected the codegen in any way.

Reply:
Argh, I missed that the `Debug.Assert(item != null)` doesn't work here. An alternative would be `_ = item!;` to pretend that it's not null, but that smells in this case, and also that's not yet implemented.
I'll push on allowing suppression on `return someBoolValue!;`.

Reply:
Another option (also smells) is to suppress on places that assign to `item` (pretend like we're never assigning maybe-null values)

---
In reply to: [422242826](https://github.com/dotnet/runtime/pull/36104#discussion_r422242826) [](ancestors = 422242826)

Reply:
Yeah, using the ! operator inappropriately early in the method strikes me as undesirable. I'll leave the pragma for now and add a code comment directing to https://github.com/dotnet/runtime/issues/36132. Once the language team settles on guidance for this we can react to it here.

Much appreciate your responsiveness on this! :)